apps/src/lib/script-editor/StageCard.jsx

```diff
@@ -0,0 +1,206 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import ReactDOM from 'react-dom';
+import {connect} from 'react-redux';
+import {borderRadius, levelTokenMargin, ControlTypes} from './constants';
+import OrderControls from './OrderControls';
+import LevelToken from './LevelToken';
+import {reorderLevel, addLevel, setStageLockable} from './editorRedux';
+
+const styles = {
+  checkbox: {
+    margin: '0 0 0 7px'
+  },
+  stageCard: {
+    fontSize: 18,
+    background: 'white',
+    border: '1px solid #ccc',
+    borderRadius: borderRadius,
+    padding: 20,
+    margin: 10
+  },
+  stageCardHeader: {
+    color: '#5b6770',
+    marginBottom: 15
+  },
+  stageLockable: {
+    fontSize: 13,
+    marginTop: 3
+  },
+  addLevel: {
+    fontSize: 14,
+    background: '#eee',
+    border: '1px solid #ddd',
+    boxShadow: 'inset 0 1px 0 0 rgba(255, 255, 255, 0.8)',
+    margin: '5px 0'
+  }
+};
+
+class StageCard extends Component {
+  static propTypes = {
+    reorderLevel: PropTypes.func.isRequired,
+    addLevel: PropTypes.func.isRequired,
+    setStageLockable: PropTypes.func.isRequired,
+    stagesCount: PropTypes.number.isRequired,
+    stage: PropTypes.object.isRequired
+  };
+
+  /**
+   * To be populated with the bounding client rect of each level token element.
+   */
+  metrics = {};
+
+  state = {
+    currentPositions: [],
+    drag: null,
+    dragHeight: null,
+    initialPageY: null,
+    initialScroll: null,
+    newPosition: null,
+    startingPositions: null
+  };
+
+  handleDragStart = (position, {pageY}) => {
+    const startingPositions = this.props.stage.levels.map(level => {
+      const metrics = this.metrics[level.position];
+      return metrics.top + metrics.height / 2;
+    });
+    this.setState({
+      drag: position,
+      dragHeight: this.metrics[position].height + levelTokenMargin,
+      initialPageY: pageY,
+      initialScroll: document.body.scrollTop,
+      newPosition: position,
+      startingPositions
+    });
```

Comment:
Pulling this locally, it looks like there's a bug with reordering in some cases:

![reorder-bug](https://user-images.githubusercontent.com/413693/60613991-d2d41580-9d80-11e9-9c6d-666d1f3dc13b.gif)

Reply:
that is weird. nice catch! will definitely want to fix that.